nflows/transforms/standard.py

```diff
@@ -41,14 +41,14 @@ def _log_scale(self):
 
     def forward(self, inputs, context=None):
         batch_size = inputs.shape[0]
-        num_dims = torch.prod(torch.tensor(inputs.shape[1:]), dtype=torch.float)
+        num_dims = torch.prod(ensure_tensor(inputs.shape[1:]), dtype=torch.float)
```

Comment:
I think it's safe to assume `inputs` will be a tensor, so no `ensure_tensor` needed. Doubt many other transforms would work with non-tensor inputs, and it's a good convention to keep.

Same in `inverse()`.

Reply:
Great, I will revert those, but then also no wrapping in `torch.tensor`?. 

Shall we keep then ensure_tensor for the other (admittedly few) cases?

Reply:
Oh, I've misread the code. We're converting the tensor shape to a tensor, not the inputs themselves -- sorry! I think we should avoid tensors all-together here and do `np.prod(inputs.shape[1:])` instead to compute `num_dims`.

Reply:
> Shall we keep then ensure_tensor for the other (admittedly few) cases?

Don't mind either way. Less code is good, but if you think it could be useful in other places at some point: feel free to leave it.

Reply:
There are only few spots where nflows depends on numpy so here a few suggestions that sidestep the dependency (and perhaps are more clear; I am not sure `num_dims` is the right name for the number of elements in a batch).

If assuming `inputs` is a tensor is safe, we could do [`inputs.numel()`](https://pytorch.org/docs/stable/torch.html#torch.numel) or, in this particular situation where we want the number of elements of all but the first dimension, `inputs[0].numel()` or even `inputs.numel()//batch_size`.  

If you still want to explicitly work on the shapes (vs. using `.numel()`), we could do this:
`batch_size, *batch_shape = inputs shape`
`batch_numel = np.prod(batch_shape)`.

Finally, as an interesting note on what Python version to require Python 3.8 offers `math.prod` (, apart from `from reduce import functools; from operator import mul; reduce(mul, inputs.shape)`, which is always possible).

Reply:
Don't want to drown you in options; just trying to get an understanding of stylistic preferences, since of some of these idioms would be applicable elsewhere ;).

Reply:
> There are only few spots where nflows depends on numpy

I don't think a dependency on such a widely used tool as `numpy` should worry us at all at this point, especially if it makes for more concise/readable code.

> I am not sure `num_dims` is the right name for the number of elements in a batch).

Agreed. `batch_numel` that you propose is better.

> `inputs[0].numel()` or even `inputs.numel()//batch_size`

I think both are more cryptic than necessary.

> `batch_size, *batch_shape = inputs.shape`
> `batch_numel = np.prod(batch_shape)`.

This works too -- very Pythonic! 🙂

> Finally, as an interesting note on what Python version to require Python 3.8 offers `math.prod`

I can't believe it has taken them this long.